src/eliater/repair.py

```diff
@@ -0,0 +1,161 @@
+        directed=[
+            ('X', 'M1'),
+            ('M1', 'M2'),
+            ('M2', 'Y'),
+        ],
+        undirected=[
+            ('X', 'Y'),
+        ],
+    )
+
+    # Generate observational data for this graph (this is a special example)
+    observational_data = generate(100)
+
+    repaired_graph = repair_network(graph, observational_data)
+
+"""
+
+import warnings
+from typing import Dict, Literal, Optional
+
+import pandas as pd
+
+from y0.algorithm.falsification import get_conditional_independencies
+from y0.dsl import Variable
+from y0.graph import NxMixedGraph
+from y0.struct import get_conditional_independence_tests
+
+__all__ = [
+    "repair_network",
+]
+
+
+def get_state_space_map(
+    data: pd.DataFrame, threshold: Optional[int] = 10
+) -> Dict[Variable, Literal["discrete", "continuous"]]:
+    """Get a dictionary from each variable to its type."""
+    column_values_unique_count = {
+        column_name: data[column_name].nunique() for column_name in data.columns
+    }
+    return {
+        Variable(column): "discrete"
+        if column_values_unique_count[column] <= threshold
+        else "continuous"
+        for column in data.columns
+    }
+
+
+def is_data_discrete(data: pd.DataFrame) -> bool:
+    """Check if all the columns in the dataframe has discrete data."""
+    variable_types = set(get_state_space_map(data=data).values())
+    return variable_types == {"discrete"}
+
+
+def is_data_continuous(data: pd.DataFrame) -> bool:
+    """Check if all the columns in the dataframe has continuous data."""
+    variable_types = set(get_state_space_map(data).values())
+    return variable_types == {"continuous"}
+
+
+CITest = Literal[
+    "pearson",
+    "chi-square",
+    "cressie_read",
+    "freeman_tuckey",
+    "g_sq",
+    "log_likelihood",
+    "modified_log_likelihood",
+    "power_divergence",
+    "neyman",
+]
+
+
+def choose_default_test(data: pd.DataFrame) -> CITest:
+    """Choose the default statistical test for testing conditional independencies based on the data."""
+    if is_data_discrete(data):
+        return "chi-square"
+    if is_data_continuous(data):
+        return "pearson"
+    raise NotImplementedError(
+        "Mixed data types are not allowed. Either all of the columns of data should be discrete / continuous."
+    )
+
+
+def repair_network(
+    graph: NxMixedGraph,
+    data: pd.DataFrame,
+    test: Optional[CITest] = None,
+    significance_level: Optional[float] = 0.01,
+) -> NxMixedGraph:
+    """Repairs the network structure.
+
+    Repairs the network structure by introducing bidirectional edges between
+    any pairs of variables when the conditional independence implied by the network
+    is not supported by the data through a statistical conditional independence test.
+
+    :param graph: an NxMixedGraph
+    :param data: observational data corresponding to the graph
+    :param test: the conditional independency test to use. Supported values are ["pearson",
```

Comment:
you don't need to explicitly document the supported values since that's part of the static typing now

Reply:
in fact, it's better you don't since they're now duplicated and can get out of sync